data race: multiple calls to maybeGossipConfigs lead to configHashes concurrent access

The following test appears to have failed:

[#2386](https://circleci.com/gh/cockroachdb/cockroach/2386):

```
level="INFO" time="2015-05-12T21:40:35Z" file="multiraft.go" line="751" msg="SoftState updated: {Lead:4294967297 RaftState:StateFollower}"
level="INFO" time="2015-05-12T21:40:35Z" file="multiraft.go" line="754" msg="HardState updated: {Term:6 Vote:0 Commit:0 XXX_unrecognized:[]}"
level="INFO" time="2015-05-12T21:40:35Z" file="multiraft.go" line="766" msg="Outgoing Message[0]: 12884901891->4294967297 MsgAppResp Term:6 Log:0/30 Rejected"
level="WARNING" time="2015-05-12T21:40:35Z" file="multiraft.go" line="905" msg="node 100000001 failed to send message to 200000002: transport.go:170: unknown peer 200000002"
level="INFO" time="2015-05-12T21:40:35Z" file="multiraft.go" line="889" msg="node 300000003: connecting to new node 100000001"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0xaebf97]

goroutine 4083 [running]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc208706c60)
	/go/src/github.com/coreos/etcd/raft/multinode.go:234 +0xc57
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:55 +0xb6

goroutine 1 [chan receive]:
testing.RunTests(0x10fb138, 0x15aff00, 0xa5, 0xa5, 0xc207ffc501)
--
goroutine 4285 [select]:
github.com/cockroachdb/cockroach/multiraft.func·010()
	/go/src/github.com/cockroachdb/cockroach/multiraft/transport.go:196 +0x1d3
created by github.com/cockroachdb/cockroach/multiraft.(*localRPCTransport).Send
	/go/src/github.com/cockroachdb/cockroach/multiraft/transport.go:204 +0x24d
FAIL	github.com/cockroachdb/cockroach/storage	3.834s
=== RUN TestBatchBasics
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:42:29Z" file="multiraft.go" line="749" msg="node 100000001: group 1 raft ready"
level="INFO" time="2015-05-12T21:42:29Z" file="multiraft.go" line="754" msg="HardState updated: {Term:6 Vote:4294967297 Commit:14 XXX_unrecognized:[]}"
level="INFO" time="2015-05-12T21:42:29Z" file="multiraft.go" line="757" msg="New Entry[0]: 6/14 EntryNormal 13dd979b2c972eb94711302ed91dda90: raft_id:1 origin_node_id:4294967297 cmd:<put:<header:<timestamp:<wall_time:0 logical:4 > cmd_id:<wall_time:1431466949297516217 random:512092722906620"
level="INFO" time="2015-05-12T21:42:29Z" file="multiraft.go" line="760" msg="Committed Entry[0]: 6/14 EntryNormal 13dd979b2c972eb94711302ed91dda90: raft_id:1 origin_node_id:4294967297 cmd:<put:<header:<timestamp:<wall_time:0 logical:4 > cmd_id:<wall_time:1431466949297516217 random:512092722906620"
==================
WARNING: DATA RACE
Write by goroutine 234:
  runtime.mapassign1()
      /usr/src/go/src/runtime/hashmap.go:383 +0x0
  github.com/cockroachdb/cockroach/storage.(*Range).maybeGossipConfigs()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:951 +0x805
  github.com/cockroachdb/cockroach/storage.(*Range).applyRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:853 +0x112c
  github.com/cockroachdb/cockroach/storage.(*Range).processRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:756 +0x3b5
  github.com/cockroachdb/cockroach/storage.func·036()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:1319 +0xce8
  github.com/cockroachdb/cockroach/util.func·001()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x7f

Previous read by goroutine 212:
  runtime.mapaccess2_fast64()
      /usr/src/go/src/runtime/hashmap_fast.go:137 +0x0
  github.com/cockroachdb/cockroach/storage.(*Range).maybeGossipConfigs()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:950 +0x6d3
  github.com/cockroachdb/cockroach/storage.(*Store).maybeGossipConfigs()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:587 +0x24c
  github.com/cockroachdb/cockroach/storage.func·029()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:529 +0x58
  github.com/cockroachdb/cockroach/util.func·001()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x7f
--
level="INFO" time="2015-05-12T21:44:20Z" file="multiraft.go" line="749" msg="node 100000001: group 1 raft ready"
level="INFO" time="2015-05-12T21:44:20Z" file="multiraft.go" line="757" msg="New Entry[0]: 6/27 EntryNormal 00000000000000007f477ed3f3ff3628: raft_id:1 origin_node_id:4294967297 cmd:<increment:<header:<timestamp:<wall_time:0 logical:75 > cmd_id:<wall_time:0 random:0 > key:\"a\" end_key:\"\" use"
level="INFO" time="2015-05-12T21:44:20Z" file="multiraft.go" line="766" msg="Outgoing Message[0]: 4294967297->12884901891 MsgApp Term:6 Log:6/26 Commit:26 Entries:[6/27 EntryNormal 00000000000000007f477ed3f3ff3628: raft_id:1 origin_node_id:4294967297 cmd:<increment:<header:<timestamp:<wall_time:0 "
level="INFO" time="2015-05-12T21:44:20Z" file="multiraft.go" line="766" msg="Outgoing Message[1]: 4294967297->8589934594 MsgApp Term:6 Log:6/26 Commit:26 Entries:[6/27 EntryNormal 00000000000000007f477ed3f3ff3628: raft_id:1 origin_node_id:4294967297 cmd:<increment:<header:<timestamp:<wall_time:0 l"
level="WARNING" time="2015-05-12T21:44:20Z" file="multiraft.go" line="905" msg="node 100000001 failed to send message to 200000002: transport.go:170: unknown peer 200000002"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0xc54259]

goroutine 4364 [running]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc2081a38c0)
	/go/src/github.com/coreos/etcd/raft/multinode.go:234 +0xfa9
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:55 +0xde

goroutine 1 [chan receive]:
testing.RunTests(0x1397ab8, 0x1889040, 0xa5, 0xa5, 0x1)
--
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:203 +0x3e4
github.com/cockroachdb/cockroach/util.func·001()
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x80
created by github.com/cockroachdb/cockroach/util.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:74 +0x133
FAIL	github.com/cockroachdb/cockroach/storage	9.753s
=== RUN TestBatchBasics
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:42:29Z" file="multiraft.go" line="749" msg="node 100000001: group 1 raft ready"
level="INFO" time="2015-05-12T21:42:29Z" file="multiraft.go" line="754" msg="HardState updated: {Term:6 Vote:4294967297 Commit:14 XXX_unrecognized:[]}"
level="INFO" time="2015-05-12T21:42:29Z" file="multiraft.go" line="757" msg="New Entry[0]: 6/14 EntryNormal 13dd979b2c972eb94711302ed91dda90: raft_id:1 origin_node_id:4294967297 cmd:<put:<header:<timestamp:<wall_time:0 logical:4 > cmd_id:<wall_time:1431466949297516217 random:512092722906620"
level="INFO" time="2015-05-12T21:42:29Z" file="multiraft.go" line="760" msg="Committed Entry[0]: 6/14 EntryNormal 13dd979b2c972eb94711302ed91dda90: raft_id:1 origin_node_id:4294967297 cmd:<put:<header:<timestamp:<wall_time:0 logical:4 > cmd_id:<wall_time:1431466949297516217 random:512092722906620"
==================
WARNING: DATA RACE
Write by goroutine 234:
  runtime.mapassign1()
      /usr/src/go/src/runtime/hashmap.go:383 +0x0
  github.com/cockroachdb/cockroach/storage.(*Range).maybeGossipConfigs()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:951 +0x805
  github.com/cockroachdb/cockroach/storage.(*Range).applyRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:853 +0x112c
  github.com/cockroachdb/cockroach/storage.(*Range).processRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:756 +0x3b5
  github.com/cockroachdb/cockroach/storage.func·036()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:1319 +0xce8
  github.com/cockroachdb/cockroach/util.func·001()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x7f

Previous read by goroutine 212:
  runtime.mapaccess2_fast64()
      /usr/src/go/src/runtime/hashmap_fast.go:137 +0x0
  github.com/cockroachdb/cockroach/storage.(*Range).maybeGossipConfigs()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:950 +0x6d3
  github.com/cockroachdb/cockroach/storage.(*Store).maybeGossipConfigs()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:587 +0x24c
  github.com/cockroachdb/cockroach/storage.func·029()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:529 +0x58
  github.com/cockroachdb/cockroach/util.func·001()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x7f
--
level="INFO" time="2015-05-12T21:44:20Z" file="multiraft.go" line="749" msg="node 100000001: group 1 raft ready"
level="INFO" time="2015-05-12T21:44:20Z" file="multiraft.go" line="757" msg="New Entry[0]: 6/27 EntryNormal 00000000000000007f477ed3f3ff3628: raft_id:1 origin_node_id:4294967297 cmd:<increment:<header:<timestamp:<wall_time:0 logical:75 > cmd_id:<wall_time:0 random:0 > key:\"a\" end_key:\"\" use"
level="INFO" time="2015-05-12T21:44:20Z" file="multiraft.go" line="766" msg="Outgoing Message[0]: 4294967297->12884901891 MsgApp Term:6 Log:6/26 Commit:26 Entries:[6/27 EntryNormal 00000000000000007f477ed3f3ff3628: raft_id:1 origin_node_id:4294967297 cmd:<increment:<header:<timestamp:<wall_time:0 "
level="INFO" time="2015-05-12T21:44:20Z" file="multiraft.go" line="766" msg="Outgoing Message[1]: 4294967297->8589934594 MsgApp Term:6 Log:6/26 Commit:26 Entries:[6/27 EntryNormal 00000000000000007f477ed3f3ff3628: raft_id:1 origin_node_id:4294967297 cmd:<increment:<header:<timestamp:<wall_time:0 l"
level="WARNING" time="2015-05-12T21:44:20Z" file="multiraft.go" line="905" msg="node 100000001 failed to send message to 200000002: transport.go:170: unknown peer 200000002"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0xc54259]

goroutine 4364 [running]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc2081a38c0)
	/go/src/github.com/coreos/etcd/raft/multinode.go:234 +0xfa9
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:55 +0xde

goroutine 1 [chan receive]:
testing.RunTests(0x1397ab8, 0x1889040, 0xa5, 0xa5, 0x1)
--
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:203 +0x3e4
github.com/cockroachdb/cockroach/util.func·001()
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x80
created by github.com/cockroachdb/cockroach/util.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:74 +0x133
FAIL	github.com/cockroachdb/cockroach/storage	9.753s
=== RUN TestBatchBasics
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
level="INFO" time="2015-05-12T21:45:06Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:40:35Z" file="multiraft.go" line="751" msg="SoftState updated: {Lead:4294967297 RaftState:StateFollower}"
level="INFO" time="2015-05-12T21:40:35Z" file="multiraft.go" line="754" msg="HardState updated: {Term:6 Vote:0 Commit:0 XXX_unrecognized:[]}"
level="INFO" time="2015-05-12T21:40:35Z" file="multiraft.go" line="766" msg="Outgoing Message[0]: 12884901891->4294967297 MsgAppResp Term:6 Log:0/30 Rejected"
level="WARNING" time="2015-05-12T21:40:35Z" file="multiraft.go" line="905" msg="node 100000001 failed to send message to 200000002: transport.go:170: unknown peer 200000002"
level="INFO" time="2015-05-12T21:40:35Z" file="multiraft.go" line="889" msg="node 300000003: connecting to new node 100000001"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0xaebf97]

goroutine 4083 [running]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc208706c60)
	/go/src/github.com/coreos/etcd/raft/multinode.go:234 +0xc57
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:55 +0xb6

goroutine 1 [chan receive]:
testing.RunTests(0x10fb138, 0x15aff00, 0xa5, 0xa5, 0xc207ffc501)
--
goroutine 4285 [select]:
github.com/cockroachdb/cockroach/multiraft.func·010()
	/go/src/github.com/cockroachdb/cockroach/multiraft/transport.go:196 +0x1d3
created by github.com/cockroachdb/cockroach/multiraft.(*localRPCTransport).Send
	/go/src/github.com/cockroachdb/cockroach/multiraft/transport.go:204 +0x24d
FAIL	github.com/cockroachdb/cockroach/storage	3.834s
=== RUN TestBatchBasics
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
level="INFO" time="2015-05-12T21:41:16Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"

```
Please assign, take a look and update the issue accordingly.

--- a/storage/range.go
+++ b/storage/range.go
@@ -765,6 +765,7 @@ func (r *Range) processRaftCommand(idKey cmdIDKey, index uint64, raftCmd proto.I
 
 // applyRaftCommand applies a raft command from the replicated log to
 // the underlying state machine (i.e. the engine).
+// The caller needs to hold the range lock.
 func (r *Range) applyRaftCommand(index uint64, originNodeID proto.RaftNodeID, args proto.Request, reply proto.Response) error {
 	if index <= 0 {
 		log.Fatalf("raft command index is <= 0")
@@ -848,7 +849,8 @@ func (r *Range) applyRaftCommand(index uint64, originNodeID proto.RaftNodeID, ar
 		switch args.(type) {
 		case *proto.PutRequest, *proto.DeleteRequest, *proto.DeleteRangeRequest:
 			if header.Key.Less(engine.KeySystemMax) {
-				r.maybeGossipConfigs(func(configPrefix proto.Key) bool {
+				// We hold the lock already.
+				r.maybeGossipConfigsLocked(func(configPrefix proto.Key) bool {
 					return bytes.HasPrefix(header.Key, configPrefix)
 				})
 			}
@@ -925,7 +927,13 @@ func (r *Range) maybeGossipFirstRange() error {
 // TODO(tschottdorf): The main reason this method does not try to get the lease
 // is that InternalLeaderLease calls it, which means that we would wind up
 // deadlocking in redirectOnOrObtainLeaderLease. Can possibly simplify.
-func (r *Range) maybeGossipConfigs(match func(configPrefix proto.Key) bool) {
+func (r *Range) maybeGossipConfigs(match func(proto.Key) bool) {
+	r.Lock()
+	defer r.Unlock()
+	r.maybeGossipConfigsLocked(match)
+}
+
+func (r *Range) maybeGossipConfigsLocked(match func(configPrefix proto.Key) bool) {
 	if r.rm.Gossip() == nil || !r.isInitialized() {
 		return
 	}

--- a/storage/range_command.go
+++ b/storage/range_command.go
@@ -760,7 +760,7 @@ func (r *Range) InternalLeaderLease(batch engine.Engine, ms *proto.MVCCStats, ar
 	}
 
 	// Gossip configs in the event this range contains config info.
-	r.maybeGossipConfigs(func(configPrefix proto.Key) bool {
+	r.maybeGossipConfigsLocked(func(configPrefix proto.Key) bool {
 		return r.ContainsKey(configPrefix)
 	})
 }

